mesonbuild/coredata.py

```diff
@@ -634,6 +634,12 @@ def set_option(self, key: OptionKey, value) -> None:
 
         if key.name == 'buildtype':
             self._set_others_from_buildtype(value)
+        elif key.name in ('wrap_mode', 'force_fallback_for'):
+            # We could have the system dependency cached for a dependency that
+            # is now forced to use subproject fallback. We probably could have
+            # more fine grained cache invalidation, but better be safe.
+            self.deps.host.clear()
+            self.deps.build.clear()
```

Comment:
Instead of clearing the cache, I would prefer if the `wrap_mode` value was stored in the key for the dependency cache in [`get_dep_identifier`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/dependencies/detect.py#L42-L68).

Reply:
I thought about that but:
- it's more work.
- it has the same effect, the whole cache is useless.
- it leaks all previous keys that are not unreachable instead of just freeing them.

Reply:
I would actually be more confortable if we invalidate the whole cache each time any option is touched. This cache is really just a small performance improvements at the cost of so many invalidation issues we already got so far...

Reply:
My main fear is that we end up with 10 different places where the cache is invalidated at some point. However, always invalidating the cache when any option changes sounds good to me.

In this case, could you also clear all the caches just in case?

Reply:
The cache is a *huge * performance improvement for mesa. Absolutely huge. invalidating the cache because *any* option changed would be a considerable regression for us.

Reply:
You often change options and rebuild? If so, then I think this PR is the right thing to do. We can always add more options into the list if we have use-cases.

Reply:
I might do something like `-Doptimization=debug` or suddenly decide "hmm, I need to build i965 as well as iris", `-Ddri-drivers=i965`. So yeah, I would be pretty annoyed if changing an option not related to dependencies suddenly caused the dependency cache to reset

Reply:
@mensinda, @dcbaker, any objection on this PR then? I refactored the cache cleaning in a method, so at least it's done from one place.